File: productButSelf.cpp
// Time Complexity :O(n) where n in the number elements in the vector
// Space Complexity : O(1)  
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No

class Solution {
public:
    vector<int> productExceptSelf(vector<int>& nums) {
        vector<int> ans;
        ans.push_back(1);
        int rP = 1;
        for(int i=1; i<nums.size();i++){
            rP *= nums[i-1];
            ans.push_back(rP);
        }
        rP = 1;
        for(int i = nums.size()-2;i>=0;i--){
            rP*=nums[i+1];
            ans[i] *=rP;
        }
        return ans;
    }
};

File: product_but_self.py
# Time Complexity :
# O (N)

# Space Complexity :
# O(1)

# Did this code successfully run on Leetcode :
#Yes

#We store the running product from the left and the right 
# We then multiply, at each position, the left product and right product to get the multiplication of everuything but the number at the current position

class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        result = [0] * len(nums)
        rp = 1
        result[0] =1

        #Left running product
        for i in range(1,len(nums)):
            rp *= nums[i-1]
            result[i] = rp
        rp = 1
        
        #Right running product and final computation by multiplying right running product and left running product
        for i in reversed(range(0,len(nums)-1)):
            rp *= nums[i+1]
            result[i] *= rp
        return result
